File: algorithms/2055-plates-between-candles/plates-between-candles.go
package problem2055

func platesBetweenCandles(s string, queries [][]int) []int {
	n := len(s)
	preSum := make([]int, n) // 前缀和
	left := make([]int, n)   // 左边第一根蜡烛所在的下标
	right := make([]int, n)  // 右边第一根蜡烛所在的下标

	sum, l := 0, -1
	for i, ch := range s {
		if ch == '*' {
			sum++
		} else {
			l = i
		}
		preSum[i] = sum
		left[i] = l
	}

	for i, r := n-1, -1; i >= 0; i-- {
		if s[i] == '|' {
			r = i
		}
		right[i] = r
	}

	ans := make([]int, len(queries))
	for i, q := range queries {
		x, y := right[q[0]], left[q[1]]
		if x >= 0 && y >= 0 && x < y {
			ans[i] = preSum[y] - preSum[x]
		}
	}
	return ans
}

File: algorithms/2055-plates-between-candles/plates-between-candles.py
"""
2055 plates between candles

蜡烛之间的盘子
"""
from typing import List

def plates_between_candles(s: str, queries: List[List[int]]) -> List[int]:
    """
    前缀和 + 预处理
    """
    n = len(s)
    pre_sum = [0] * n
    left = [0] * n
    right = [0] * n

    tmp_pre_sum, l = 0, -1
    for i, ch in enumerate(s):
        if ch == '*':
            tmp_pre_sum += 1
        else:
            l = i
        pre_sum[i] = tmp_pre_sum
        left[i] = l
    
    r = n - 1
    for i in range(n - 1, -1, - 1):
        if s[i] == '|':
            r = i
        right[i] = r

    ans = [0] * len(queries)
    for i, q in enumerate(queries):
        x, y = right[q[0]], left[q[1]]
        if x >= 0 and y >= 0 and x < y:
            ans[i] = pre_sum[y] - pre_sum[x]
    return ans
